cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/LogParserTransform.java

```diff
@@ -0,0 +1,183 @@
+      "(\\d{3}) (\\p{Print}+) ([-\"\\d]+) ([-\"\\d]+) ([\\d]+) ([-\"\\d]+) \"(\\p{Print}+)\" \"([^\"]+)\" " +
+      //  version id
+      "(\\p{Print}+)");
+  private static final Pattern REQUEST_PAGE_PATTERN = Pattern.compile("(\\S+)\\s(\\S+).*");
+  private static final int[] S3_INDICES = {9, 3, 4, 17};
+  private static final int[] CLF_INDICES = {5, 4, 1, 9};
+  private static final int S3_REGEX_LENGTH = 18;
+  private static final int CLF_REGEX_LENGTH = 9;
+  private final LogParserConfig config;
+
+  public LogParserTransform(LogParserConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    String log;
+    if ("string".equals(config.inputType)) {
+      log = input.get(config.inputName);
+    } else if ("bytes".equals(config.inputType)) {
+      log = Bytes.toString((ByteBuffer) input.get(config.inputName));
+    } else {
+      LOG.error("Unsupported inputType");
+      return;
+    }
+
+    if (log == null) {
+        LOG.error("Couldn't read schema");
+        return;
+    }
+
+    if ("S3".equals(config.logFormat)) {
+      Matcher logMatcher = S3_LOG_PATTERN.matcher(log);
+      if (!logMatcher.matches() || logMatcher.groupCount() < S3_REGEX_LENGTH) {
+        LOG.error("Couldn't parse log " + log);
+        return;
+      }
+      parseRequest(emitter, logMatcher, S3_INDICES);
+    } else if ("CLF".equals(config.logFormat)) {
+      Matcher logMatcher = CLF_LOG_PATTERN.matcher(log);
+      if (!logMatcher.matches() || logMatcher.groupCount() < CLF_REGEX_LENGTH) {
+        LOG.error("Couldn't parse log " + log);
+        return;
+      }
+      parseRequest(emitter, logMatcher, CLF_INDICES);
+    } else {
+      LOG.error("Log format not currently supported.");
+      return;
+    }
+  }
+
+  /**
+   * Parses a request for the URI, IP, Browser, Device, and Time
+   * @param emitter the emitter to emit parsed information to
+   * @param logMatcher the regex matcher to use
+   * @param indices array of indices that define what position in the regex the fields are,
+   *                 in the order of Request, Time, IP, and User Agent
+   */
+  private void parseRequest(Emitter<StructuredRecord> emitter, Matcher logMatcher, int[] indices) {
+    String request = logMatcher.group(indices[0]);
+    long ts = System.currentTimeMillis();
+    Matcher requestMatcher = REQUEST_PAGE_PATTERN.matcher(request);
+    if (!requestMatcher.matches() || requestMatcher.groupCount() < 2) {
+      LOG.error("Couldn't parse uri " + request);
+      return;
+    }
+
+    String uri = requestMatcher.group(2);
+    try {
+      //Needs to be local because SimpleDateFormat is not thread safe
+      SimpleDateFormat sdf = new SimpleDateFormat("dd/MMM/yyyy:HH:mm:ss Z");
+      ts = sdf.parse(logMatcher.group(indices[1])).getTime();
+    } catch (ParseException e) {
+      LOG.info("Couldn't parse time, using ts of System.currentTimeMillis()");
```

Comment:
debug. 'Couldn't parse time from the input record, using current timestamp instead.'?